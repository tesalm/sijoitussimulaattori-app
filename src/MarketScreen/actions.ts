import { Dispatch } from 'redux';

import { stockHistoryApiRequest, stockIntraApiRequest, stockListApiRequest, stockMetaApiRequest } from '../utils/api';
import { HistoryDataQuote, IntradayQuote, Stock, StockMetadata } from './reducer';

export enum ActionType {
  RequestStocksBegin = '[Stocks] API Request',
  RequestStocksSuccess = '[Stocks] API Success',
  RequestStocksFailure = '[Stocks] API Failure',
  SaveAsCurrentSymbol = '[Stocks] Save Active Stock Symbol',
  UpdateStockData = '[Stocks] Update Stock Data',
  GetStockMetadataBegin = '[Stocks] StockMetadata begin',
  GetStockMetadataSuccess = '[Stocks] StockMetadata success',
  GetStockMetadataFailure = '[Stocks] StockMetadata failure',
  GetIntradayBegin = '[Stocks] Intraday begin',
  GetIntradaySuccess = '[Stocks] Intraday success',
  GetIntradayFailure = '[Stocks] Intraday failure',
  RefreshIntradayBegin = '[Stocks] Intraday begin refresh',
  RefreshIntradaySuccess = '[Stocks] Intraday refresh success',
  RefreshIntradayFailure = '[Stocks] Intraday refresh failure',
  GetHistoryBegin = '[Stocks] History begin',
  GetHistorySuccess = '[Stocks] History success',
  GetHistoryFailure = '[Stocks] History failure',
}

export type StockAction =
  | RequestStocksBegin
  | RequestStocksSuccess
  | RequestStocksFailure
  | SaveAsCurrentSymbol
  | GetStockMetadataBegin
  | GetStockMetadataSuccess
  | GetStockMetadataFailure
  | GetIntradayBegin
  | GetIntradaySuccess
  | GetIntradayFailure
  | RefreshIntradayBegin
  | RefreshIntradaySuccess
  | RefreshIntradayFailure
  | GetHistoryBegin
  | GetHistorySuccess
  | GetHistoryFailure;

export class RequestStocksBegin {
  readonly type = ActionType.RequestStocksBegin;
  constructor() {
    return { type: this.type };
  }
}

export class RequestStocksSuccess {
  readonly type = ActionType.RequestStocksSuccess;
  constructor(public stocks: Array<Stock>) {
    return { type: this.type, stocks };
  }
}

export class RequestStocksFailure {
  readonly type = ActionType.RequestStocksFailure;
  constructor(public error: Error) {
    return { type: this.type, error };
  }
}

export class SaveAsCurrentSymbol {
  readonly type = ActionType.SaveAsCurrentSymbol;
  constructor(public symbol: string) {
    return { type: this.type, symbol };
  }
}

export class GetStockMetadataBegin {
  readonly type = ActionType.GetStockMetadataBegin;
  constructor(public symbol: string) {
    return { type: this.type, symbol };
  }
}

export class GetStockMetadataSuccess {
  readonly type = ActionType.GetStockMetadataSuccess;
  constructor(
    public symbol: string,
    public metaData: StockMetadata,
    public fetchTime: Date
  ) {
    return { type: this.type, symbol, metaData, fetchTime };
  }
}

export class GetStockMetadataFailure {
  readonly type = ActionType.GetStockMetadataFailure;
  constructor(public symbol: string, public error: Error) {
    return { type: this.type, symbol, error };
  }
}

export class GetIntradayBegin {
  readonly type = ActionType.GetIntradayBegin;
  constructor(public symbol: string) {
    return { type: this.type, symbol };
  }
}

export class GetIntradaySuccess {
  readonly type = ActionType.GetIntradaySuccess;
  constructor(
    public symbol: string,
    public intraData: IntradayQuote[],
    public fetchTime: Date
  ) {
    return { type: this.type, symbol, intraData, fetchTime };
  }
}

export class GetIntradayFailure {
  readonly type = ActionType.GetIntradayFailure;
  constructor(public symbol: string, public error: Error) {
    return { type: this.type, symbol, error };
  }
}

export class RefreshIntradayBegin {
  readonly type = ActionType.RefreshIntradayBegin;
  constructor(public symbol: string) {
    return { type: this.type, symbol };
  }
}

export class RefreshIntradaySuccess {
  readonly type = ActionType.RefreshIntradaySuccess;
  constructor(
    public symbol: string,
    public intraData: IntradayQuote[],
    public fetchTime: Date
  ) {
    return { type: this.type, symbol, intraData, fetchTime };
  }
}

export class RefreshIntradayFailure {
  readonly type = ActionType.RefreshIntradayFailure;
  constructor(public symbol: string, public error: Error) {
    return { type: this.type, symbol, error };
  }
}

export class GetHistoryBegin {
  readonly type = ActionType.GetHistoryBegin;
  constructor(public symbol: string) {
    return { type: this.type, symbol };
  }
}

export class GetHistorySuccess {
  readonly type = ActionType.GetHistorySuccess;
  constructor(
    public symbol: string,
    public historyData: HistoryDataQuote[],
    public fetchTime: Date
  ) {
    return { type: this.type, symbol, historyData, fetchTime };
  }
}
export class GetHistoryFailure {
  readonly type = ActionType.GetHistoryFailure;
  constructor(public symbol: string, public error: Error) {
    return { type: this.type, symbol, error };
  }
}

// API-request for getting stocks.
const getStocks = () => async (dispatch: Dispatch<StockAction>) => {
  dispatch(new RequestStocksBegin());
  try {
    const data = await stockListApiRequest();
    data.forEach((stock) => {
      stock.stockInfo = {
        metaLoading: false,
        intraLoading: false,
        historyLoading: false,
        refreshing: false,
      };
    });
    dispatch(new RequestStocksSuccess(data));
  } catch (error) {
    dispatch(new RequestStocksFailure(error));
  }
};

// Saves the symbol of a stock that is shown to the user.
const saveAsCurrentStockSymbol = (symbol: string) => async (
  dispatch: Dispatch<StockAction>
) => {
  dispatch(new SaveAsCurrentSymbol(symbol));
};

// API-request for getting metadata for single stock.
const getStockMetadata = (stock: Stock, symbol: string) => async (
  dispatch: Dispatch<StockAction>
) => {
  // Metadata is being fetched once a day, if not undefined.
  const currentTime = new Date();
  if (
    stock.stockInfo === undefined ||
    stock.stockInfo.stockMetadata === undefined ||
    onceADayRefreshrateDated(
      currentTime,
      stock.stockInfo.stockMetadata.fetchTime
    )
  ) {
    dispatch(new GetStockMetadataBegin(symbol));
    try {
      const meta = await stockMetaApiRequest(symbol);
      dispatch(new GetStockMetadataSuccess(symbol, meta, currentTime));
    } catch (error) {
      dispatch(new GetStockMetadataFailure(symbol, error));
    }
  }
};

// API-request for getting intraday-data.
const getIntraday = (stock: Stock, symbol: string) => async (
  dispatch: Dispatch<StockAction>
) => {
  // Intraday is fetched several times a day.
  const currentTime = new Date();
  if (
    stock.stockInfo === undefined ||
    stock.stockInfo.intraday === undefined ||
    refreshrateDated(currentTime, stock.stockInfo.intraday.fetchTime)
  ) {
    dispatch(new GetIntradayBegin(symbol));
    try {
      const intraData = await stockIntraApiRequest(symbol);
      dispatch(new GetIntradaySuccess(symbol, intraData, currentTime));
    } catch (error) {
      dispatch(new GetIntradayFailure(symbol, error));
    }
  }
};

// API-request for getting intraday-data. Called only when user refreshes
// the screen manually (for example swiping down).
const refreshIntraday = (stock: Stock, symbol: string) => async (
  dispatch: Dispatch<StockAction>
) => {
  const currentTime = new Date();
  if (
    stock.stockInfo === undefined ||
    stock.stockInfo.intraday === undefined ||
    refreshrateDated(currentTime, stock.stockInfo.intraday.fetchTime)
  ) {
    dispatch(new RefreshIntradayBegin(symbol));
    try {
      const intraData = await stockIntraApiRequest(symbol);
      dispatch(new RefreshIntradaySuccess(symbol, intraData, currentTime));
    } catch (error) {
      dispatch(new RefreshIntradayFailure(symbol, error));
    }
  }
};

// API-request for getting historydata
const getHistory = (stock: Stock, symbol: string) => async (
  dispatch: Dispatch<StockAction>
) => {
  const currentTime = new Date();

  // Historydata is being refreshed once a day, if not undefined.
  if (
    stock.stockInfo === undefined ||
    stock.stockInfo.historyData === undefined ||
    onceADayRefreshrateDated(currentTime, stock.stockInfo.historyData.fetchTime)
  ) {
    dispatch(new GetHistoryBegin(symbol));
    try {
      const historyData = await stockHistoryApiRequest(symbol);
      dispatch(new GetHistorySuccess(symbol, historyData, currentTime));
    } catch (error) {
      dispatch(new GetHistoryFailure(symbol, error));
    }
  }
};

// Helper functions:
// Checks if metadata and historydata should be updated (they should be updated once a day)
function onceADayRefreshrateDated(curTime: Date, fetchTime: Date): boolean {
  // TODO: Set updateTime (once a day). Now 00:30.00
  const updateTimeH = 0;
  const updateTimeM = 30;
  const updateTimeS = 0;

  // If curTime > 00:30.00
  if (
    curTime.getHours() >= updateTimeH &&
    curTime.getMinutes() >= updateTimeM &&
    curTime.getSeconds() >= updateTimeS
  ) {
    // If data has been fetched the day before or fetchTime < updateTime, refresh data
    if (
      curTime.getDate() != fetchTime.getDate() ||
      (fetchTime.getHours() <= updateTimeH &&
        fetchTime.getMinutes() <= updateTimeM &&
        fetchTime.getSeconds() <= updateTimeS)
    ) {
      return true;
    }
  }
  return false;
}

// Checks if intraday should be updated (intraday is updated many times a day)
function refreshrateDated(curTime: Date, fetchTime: Date): boolean {
  const curTime_ms = curTime.getTime();
  const intraTime_ms = fetchTime.getTime();
  // TODO: Set refreshrate. Now 5 minutes.
  if (curTime_ms - intraTime_ms > 1000 * 60 * 5) {
    return true;
  }
  return false;
}

export {
  getStocks,
  saveAsCurrentStockSymbol,
  getStockMetadata,
  getIntraday,
  refreshIntraday,
  getHistory,
};
