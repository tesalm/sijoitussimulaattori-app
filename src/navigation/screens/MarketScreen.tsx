import React from 'react';
import { Button, Text, View } from 'react-native';
import { NavigationScreenProps } from 'react-navigation';

import t from '../../assets/i18n';
import { RouteName } from '../routes';

export default class MarketScreen extends React.Component<
  NavigationScreenProps
> {
  static navigationOptions = { title: t('MarketPage.Title') };
  render() {
    return (
      <View style={{ flex: 1, alignItems: 'center', justifyContent: 'center' }}>
        <Text>{t('MarketPage.PlaceholderText')}</Text>
        <View style={{ padding: 10 }}>
          <Button
            title={t('MarketPage.GoToCommissionsButtonLabel')}
            onPress={() =>
              this.props.navigation.navigate(RouteName.Commissions)
            }
          />
        </View>
        <View style={{ padding: 10 }}>
          <Button
            title={t('MarketPage.GoToHomePageButtonLabel')}
            onPress={() => this.props.navigation.navigate(RouteName.Home)}
          />
        </View>
      </View>
    );
  }
}
