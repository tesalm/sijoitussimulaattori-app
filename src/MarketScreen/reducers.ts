import { ActionType, StocksAction } from './actions';
import { Stock } from '../redux/reducers';

export interface StocksListing {
  stocks: Array<Stock>;
  loading: boolean;
  error?: Error;
}

const initialState: StocksListing = {
  stocks: [],
  loading: false,
  error: undefined,
};

export const stocksListingReducer = (
  state: StocksListing = initialState,
  action: StocksAction
): StocksListing => {
  switch (action.type) {
    case ActionType.RequestStocksBegin:
      return { ...state, loading: true, error: undefined };
    case ActionType.RequestStocksSuccess:
      return {
        ...state,
        stocks: action.stocks,
        loading: false,
        error: undefined,
      };
    case ActionType.RequestStocksFailure:
      return { ...state, loading: false, error: action.error };
    default:
      return state;
  }
};
