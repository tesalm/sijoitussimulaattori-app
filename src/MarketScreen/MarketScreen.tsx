import React from 'react';
import { ActivityIndicator, FlatList, RefreshControl, Text, ToastAndroid, View } from 'react-native';
import { ListItem, SearchBar } from 'react-native-elements';
import { NavigationScreenProps } from 'react-navigation';
import { connect } from 'react-redux';
import { bindActionCreators, Dispatch } from 'redux';

import { Colors } from '../App/colors';
import { t } from '../assets/i18n';
import { RootState } from '../redux/reducers';
import { formatCurrency, formatRevenue, revenueColor } from '../util/general';
import { getStocks, saveAsCurrentStockSymbol } from './actions';
import { Stock } from './reducer';
import { StockStyles } from './styles';

export interface StockProps {
  stocks: Array<Stock>;
  loading: boolean;
  refreshing: boolean;
  error?: Error;
  getAllStocks: typeof getStocks;
  saveAsCurrentSymbol: typeof saveAsCurrentStockSymbol;
}

interface StockState {}

type StockPropsWithNavigation = StockProps & NavigationScreenProps;

export class MarketScreen extends React.Component<
  StockPropsWithNavigation,
  StockState
> {
  constructor(props: StockPropsWithNavigation) {
    super(props);
  }
  static navigationOptions = { title: t('MarketPage.Title') };

  componentDidMount() {
    //Dispatch the actions
    this.props.getAllStocks();
  }

  renderHeader = (): JSX.Element => {
    return (
      <SearchBar
        lightTheme
        round
        placeholder={t('ListStockPage.SearcBarPlaceholder')} //TODO: search bar functionality
        autoCorrect={false}
      />
    );
  };

  //Every other listitem has gray background
  listBackgroundColor = (index: number): typeof StockStyles.greyContainer => {
    return index % 2 ? StockStyles.greyContainer : StockStyles.whiteContainer;
  };

  refresh = () => {
    this.props.getAllStocks();
  };

  stockPressed = (symbol: string) => {
    if (this.props.refreshing) {
      // TODO: Format toast message to user.
      ToastAndroid.show(
        'Wait, stock-list is being refreshed.',
        ToastAndroid.SHORT
      );
    } else {
      this.props.saveAsCurrentSymbol(symbol);
      this.props.navigation.navigate('SingleStock');
    }
  };

  render() {
    const { stocks, loading, refreshing, error } = this.props;
    if (error) {
      //TODO: Format the error message to user
      return <Text>Error! {error.message} </Text>;
    }

    // Loading stocks without swiping
    if (loading) {
      return (
        <View style={StockStyles.loadingView}>
          <ActivityIndicator size="large" />
        </View>
      );
    }

    return (
      <FlatList
        refreshControl={
          <RefreshControl
            refreshing={refreshing}
            onRefresh={this.refresh}
            colors={[Colors.baseColor]}
          />
        }
        data={stocks}
        keyExtractor={(item) => item.symbol}
        renderItem={({ item, index }) => (
          <ListItem
            onPress={() => this.stockPressed(item.symbol)}
            containerStyle={this.listBackgroundColor(index)}
            title={item.name}
            titleStyle={StockStyles.titleStyle}
            rightTitle={
              <View style={StockStyles.rightTitleView}>
                <Text style={StockStyles.revenueText}>
                  {t('ListStockPage.RevenueText')}
                </Text>
                <Text style={revenueColor(item.revenue)}>
                  {formatRevenue(item.revenue)}
                </Text>
              </View>
            }
            subtitle={
              <View style={StockStyles.subtitleView}>
                <Text style={StockStyles.lastSaleText}>
                  {t('ListStockPage.LastSaleText')}
                </Text>
                <Text style={StockStyles.lastSaleValue}>
                  {formatCurrency(item.close, item.currency)}
                </Text>
              </View>
            }
          />
        )}
        ListHeaderComponent={this.renderHeader}
      />
    );
  }
}
const mapStateToProps = (state: RootState) => ({
  stocks: state.stocksListing.stocks,
  loading: state.stocksListing.loading,
  refreshing: state.stocksListing.refreshing,
  error: state.stocksListing.error,
});

const mapDispatchToProps = (dispatch: Dispatch) =>
  bindActionCreators(
    {
      getAllStocks: getStocks,
      saveAsCurrentSymbol: saveAsCurrentStockSymbol,
    },
    dispatch
  );
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(MarketScreen);
