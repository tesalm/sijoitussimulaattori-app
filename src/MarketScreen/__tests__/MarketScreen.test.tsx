import * as React from 'react';
import renderer from 'react-test-renderer';

import { MarketScreen, StockProps } from '../MarketScreen';

describe('stock-list tests', () => {
  //Mock for navigation.
  const navigationMock: any = {};

  const defaultStockProps: StockProps = {
    stocks: [],
    loading: false,
    refreshing: false,
    error: undefined,
    getAllStocks: jest.fn(),
    saveAsCurrentSymbol: jest.fn(),
  };

  const loadingStockProps: StockProps = {
    stocks: [],
    loading: true,
    refreshing: false,
    error: undefined,
    getAllStocks: jest.fn(),
    saveAsCurrentSymbol: jest.fn(),
  };

  const refreshingStockProps: StockProps = {
    stocks: [],
    loading: false,
    refreshing: true,
    error: undefined,
    getAllStocks: jest.fn(),
    saveAsCurrentSymbol: jest.fn(),
  };

  const errorStockProps: StockProps = {
    stocks: [],
    loading: false,
    refreshing: false,
    error: { name: 'Network Error', message: 'Network connection failed' },
    getAllStocks: jest.fn(),
    saveAsCurrentSymbol: jest.fn(),
  };

  const stocksStockProps: StockProps = {
    stocks: [
      {
        symbol: 'APL',
        name: 'Apple',
        high: 10.28,
        low: 9.99,
        revenue: -0.035,
        close: 14.57,
        currency: 'USD',
        stockInfo: {
          metaLoading: false,
          intraLoading: false,
          historyLoading: false,
          refreshing: false,
        },
      },
      {
        symbol: 'APLL',
        name: 'Apple',
        high: 10.28,
        low: 9.99,
        revenue: -0.035,
        close: 14.57,
        currency: 'USD',
        stockInfo: {
          metaLoading: false,
          intraLoading: false,
          historyLoading: false,
          refreshing: false,
        },
      },
    ],
    loading: false,
    refreshing: false,
    error: undefined,
    getAllStocks: jest.fn(),
    saveAsCurrentSymbol: jest.fn(),
  };

  it('renders correctly', async () => {
    const component = renderer
      .create(<MarketScreen {...defaultStockProps} {...navigationMock} />)
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  it('renders correctly with loading', async () => {
    const component = renderer
      .create(<MarketScreen {...loadingStockProps} {...navigationMock} />)
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  it('renders correctly with refreshing', async () => {
    const component = renderer
      .create(<MarketScreen {...refreshingStockProps} {...navigationMock} />)
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  it('renders correctly with error', async () => {
    const component = renderer
      .create(<MarketScreen {...errorStockProps} {...navigationMock} />)
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  it('renders correctly with stocks', async () => {
    const component = renderer
      .create(<MarketScreen {...stocksStockProps} {...navigationMock} />)
      .toJSON();
    expect(component).toMatchSnapshot();
  });
});
