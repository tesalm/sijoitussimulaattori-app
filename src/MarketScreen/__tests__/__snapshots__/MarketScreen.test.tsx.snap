// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hello World renders correctly 1`] = `
<RCTScrollView
  ListHeaderComponent={[Function]}
  data={Array []}
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
    >
      <View
        style={
          Object {
            "backgroundColor": "#e1e8ee",
            "borderBottomColor": "#e1e1e1",
            "borderBottomWidth": 1,
            "borderTopColor": "#e1e1e1",
            "borderTopWidth": 1,
            "padding": 8,
          }
        }
      >
        <View
          style={
            Object {
              "width": "100%",
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#bdc6cf",
                "borderBottomWidth": 0,
                "borderColor": "#86939e",
                "borderRadius": 15,
                "flexDirection": "row",
                "height": 30,
                "overflow": "hidden",
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                ],
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "height": 40,
                  "justifyContent": "center",
                  "marginLeft": 8,
                }
              }
            >
              <View>
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "transparent",
                      "justifyContent": "center",
                    }
                  }
                  theme={
                    Object {
                      "colors": Object {
                        "divider": "#bcbbc1",
                        "error": "#ff190c",
                        "grey0": "#393e42",
                        "grey1": "#43484d",
                        "grey2": "#5e6977",
                        "grey3": "#86939e",
                        "grey4": "#bdc6cf",
                        "grey5": "#e1e8ee",
                        "greyOutline": "#bbb",
                        "primary": "#2089dc",
                        "searchBg": "#303337",
                        "secondary": "#8F0CE8",
                      },
                    }
                  }
                  underlayColor="white"
                >
                  <Text
                    allowFontScaling={false}
                    style={
                      Array [
                        Object {
                          "color": "#86939e",
                          "fontSize": 18,
                        },
                        Object {
                          "backgroundColor": "transparent",
                        },
                        Object {
                          "fontFamily": "Material Design Icons",
                          "fontStyle": "normal",
                          "fontWeight": "normal",
                        },
                        Object {},
                      ]
                    }
                    testID="iconIcon"
                  >
                    Ôçâ
                  </Text>
                </View>
              </View>
            </View>
            <TextInput
              allowFontScaling={true}
              autoCorrect={false}
              onBlur={[Function]}
              onChangeText={[Function]}
              onClear={[Function]}
              onFocus={[Function]}
              placeholder="Search for stocks"
              placeholderTextColor="#86939e"
              platform="default"
              style={
                Object {
                  "alignSelf": "center",
                  "color": "#86939e",
                  "flex": 1,
                  "fontSize": 18,
                  "height": 40,
                  "marginLeft": 10,
                }
              }
              testID="searchInput"
              underlineColorAndroid="transparent"
            />
            <View
              style={
                Object {
                  "alignItems": "center",
                  "height": 40,
                  "justifyContent": "center",
                  "marginLeft": 15,
                  "marginRight": 8,
                }
              }
            >
              <View
                style={
                  Object {
                    "flexDirection": "row",
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`Hello World renders correctly with error 1`] = `
<Text>
  Error! 
  Network connection failed
   
</Text>
`;

exports[`Hello World renders correctly with loading 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="#999999"
    hidesWhenStopped={true}
    size="large"
  />
</View>
`;

exports[`Hello World renders correctly with stocks 1`] = `
<RCTScrollView
  ListHeaderComponent={[Function]}
  data={
    Array [
      Object {
        "buy": 10.26,
        "close": 14.57,
        "high": 10.28,
        "low": 9.99,
        "marketValue": 10.26,
        "name": "Apple",
        "revenue": -0.035,
        "sell": 10.26,
        "symbol": "APL",
      },
      Object {
        "buy": 10.26,
        "close": 14.57,
        "high": 10.28,
        "low": 9.99,
        "marketValue": 10.26,
        "name": "Apple",
        "revenue": -0.035,
        "sell": 10.26,
        "symbol": "APLL",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
    >
      <View
        style={
          Object {
            "backgroundColor": "#e1e8ee",
            "borderBottomColor": "#e1e1e1",
            "borderBottomWidth": 1,
            "borderTopColor": "#e1e1e1",
            "borderTopWidth": 1,
            "padding": 8,
          }
        }
      >
        <View
          style={
            Object {
              "width": "100%",
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#bdc6cf",
                "borderBottomWidth": 0,
                "borderColor": "#86939e",
                "borderRadius": 15,
                "flexDirection": "row",
                "height": 30,
                "overflow": "hidden",
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                ],
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "height": 40,
                  "justifyContent": "center",
                  "marginLeft": 8,
                }
              }
            >
              <View>
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "transparent",
                      "justifyContent": "center",
                    }
                  }
                  theme={
                    Object {
                      "colors": Object {
                        "divider": "#bcbbc1",
                        "error": "#ff190c",
                        "grey0": "#393e42",
                        "grey1": "#43484d",
                        "grey2": "#5e6977",
                        "grey3": "#86939e",
                        "grey4": "#bdc6cf",
                        "grey5": "#e1e8ee",
                        "greyOutline": "#bbb",
                        "primary": "#2089dc",
                        "searchBg": "#303337",
                        "secondary": "#8F0CE8",
                      },
                    }
                  }
                  underlayColor="white"
                >
                  <Text
                    allowFontScaling={false}
                    style={
                      Array [
                        Object {
                          "color": "#86939e",
                          "fontSize": 18,
                        },
                        Object {
                          "backgroundColor": "transparent",
                        },
                        Object {
                          "fontFamily": "Material Design Icons",
                          "fontStyle": "normal",
                          "fontWeight": "normal",
                        },
                        Object {},
                      ]
                    }
                    testID="iconIcon"
                  >
                    Ôçâ
                  </Text>
                </View>
              </View>
            </View>
            <TextInput
              allowFontScaling={true}
              autoCorrect={false}
              onBlur={[Function]}
              onChangeText={[Function]}
              onClear={[Function]}
              onFocus={[Function]}
              placeholder="Search for stocks"
              placeholderTextColor="#86939e"
              platform="default"
              style={
                Object {
                  "alignSelf": "center",
                  "color": "#86939e",
                  "flex": 1,
                  "fontSize": 18,
                  "height": 40,
                  "marginLeft": 10,
                }
              }
              testID="searchInput"
              underlineColorAndroid="transparent"
            />
            <View
              style={
                Object {
                  "alignItems": "center",
                  "height": 40,
                  "justifyContent": "center",
                  "marginLeft": 15,
                  "marginRight": 8,
                }
              }
            >
              <View
                style={
                  Object {
                    "flexDirection": "row",
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#F0F0F0",
              "borderBottomWidth": 0,
              "borderColor": "#bcbbc1",
              "flexDirection": "row",
              "height": 147.13235294117646,
              "padding": 14,
            }
          }
        >
          <View
            style={
              Object {
                "flex": 1,
                "justifyContent": "center",
              }
            }
          >
            <Text
              style={
                Object {
                  "backgroundColor": "transparent",
                  "color": "#004D40",
                  "fontSize": 16,
                  "fontWeight": "bold",
                  "marginLeft": 17.142857142857142,
                  "marginTop": 15.694117647058823,
                }
              }
              testID="listItemTitle"
              theme={
                Object {
                  "colors": Object {
                    "divider": "#bcbbc1",
                    "error": "#ff190c",
                    "grey0": "#393e42",
                    "grey1": "#43484d",
                    "grey2": "#5e6977",
                    "grey3": "#86939e",
                    "grey4": "#bdc6cf",
                    "grey5": "#e1e8ee",
                    "greyOutline": "#bbb",
                    "primary": "#2089dc",
                    "searchBg": "#303337",
                    "secondary": "#8F0CE8",
                  },
                }
              }
            >
              Apple
            </Text>
            <View
              style={
                Object {
                  "flexDirection": "column",
                  "marginLeft": 17.142857142857142,
                  "marginTop": 23.541176470588233,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#004D40",
                    "fontSize": 12,
                    "marginBottom": 0,
                    "textAlign": "justify",
                  }
                }
              >
                Last sale
              </Text>
              <Text
                style={
                  Object {
                    "color": "#004D40",
                    "fontSize": 14,
                    "fontWeight": "bold",
                    "marginTop": -3.9235294117647057,
                    "textAlign": "justify",
                  }
                }
              >
                14.57 $
              </Text>
            </View>
          </View>
          <View
            width={16}
          />
          <View
            style={
              Object {
                "alignItems": "flex-end",
                "flex": 0.5,
                "justifyContent": "center",
              }
            }
          >
            <View
              style={
                Object {
                  "flexDirection": "column",
                  "marginLeft": 17.142857142857142,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#004D40",
                    "fontSize": 12,
                  }
                }
              >
                Revenue in 24h
              </Text>
              <Text
                style={
                  Object {
                    "color": "#e20f00",
                    "fontSize": 18,
                    "fontWeight": "bold",
                    "textAlign": "center",
                  }
                }
              >
                -3.50 %
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "white",
              "borderBottomWidth": 0,
              "borderColor": "#bcbbc1",
              "flexDirection": "row",
              "height": 147.13235294117646,
              "padding": 14,
            }
          }
        >
          <View
            style={
              Object {
                "flex": 1,
                "justifyContent": "center",
              }
            }
          >
            <Text
              style={
                Object {
                  "backgroundColor": "transparent",
                  "color": "#004D40",
                  "fontSize": 16,
                  "fontWeight": "bold",
                  "marginLeft": 17.142857142857142,
                  "marginTop": 15.694117647058823,
                }
              }
              testID="listItemTitle"
              theme={
                Object {
                  "colors": Object {
                    "divider": "#bcbbc1",
                    "error": "#ff190c",
                    "grey0": "#393e42",
                    "grey1": "#43484d",
                    "grey2": "#5e6977",
                    "grey3": "#86939e",
                    "grey4": "#bdc6cf",
                    "grey5": "#e1e8ee",
                    "greyOutline": "#bbb",
                    "primary": "#2089dc",
                    "searchBg": "#303337",
                    "secondary": "#8F0CE8",
                  },
                }
              }
            >
              Apple
            </Text>
            <View
              style={
                Object {
                  "flexDirection": "column",
                  "marginLeft": 17.142857142857142,
                  "marginTop": 23.541176470588233,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#004D40",
                    "fontSize": 12,
                    "marginBottom": 0,
                    "textAlign": "justify",
                  }
                }
              >
                Last sale
              </Text>
              <Text
                style={
                  Object {
                    "color": "#004D40",
                    "fontSize": 14,
                    "fontWeight": "bold",
                    "marginTop": -3.9235294117647057,
                    "textAlign": "justify",
                  }
                }
              >
                14.57 $
              </Text>
            </View>
          </View>
          <View
            width={16}
          />
          <View
            style={
              Object {
                "alignItems": "flex-end",
                "flex": 0.5,
                "justifyContent": "center",
              }
            }
          >
            <View
              style={
                Object {
                  "flexDirection": "column",
                  "marginLeft": 17.142857142857142,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#004D40",
                    "fontSize": 12,
                  }
                }
              >
                Revenue in 24h
              </Text>
              <Text
                style={
                  Object {
                    "color": "#e20f00",
                    "fontSize": 18,
                    "fontWeight": "bold",
                    "textAlign": "center",
                  }
                }
              >
                -3.50 %
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
